rules/diff_test.bzl

```diff
@@ -0,0 +1,145 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A test rule that compares two binary files.
+
+The rule uses a Bash command (diff) on Linux/macOS/non-Windows, and a cmd.exe
+command (fc.exe) on Windows (no Bash is required).
+"""
+
+def _runfiles_path(f):
+    if f.root.path:
+        return f.path[len(f.root.path) + 1:]  # generated file
+    else:
+        return f.path  # source file
+
+def _diff_test_impl(ctx):
+    if ctx.attr.is_windows:
+        test_bin = ctx.actions.declare_file(ctx.label.name + "-test.bat")
+        ctx.actions.write(
+            output = test_bin,
+            content = r"""@echo off
```

Comment:
I find it a little awkward to have this much content written as a string literal in Starlark. Could we move it to a bat file and use that as an action input?

Reply:
I tried, but couldn't: I created templates under `//rules/private` and added a label attribute to `_diff_test` that `select()`s from them:
```
    diff_script_template = select({
        "@bazel_tools//src/conditions:host_windows": "//rules/private:diff_test_tmpl.bat",
        "//conditions:default": "//rules/private:diff_test_tmpl.sh",
    }),
```

Unfortunately the rule can't find the templates when the rule itself is in an external repo.

If you don't mind, I'd leave the ugly strings here; that approach works.

I edited a test to pull the rule from an external repo.

Reply:
I don't understand the issue you're running into ...
1. It's not necessary to select on the diff script -- you can depend on the scripts for both windows and non-windows, and only actually use the script you need based on the value of is_windows.
2. Depending on a script "tool" dependency is a pretty normal pattern for rules. If there's a serious problem with adopting this pattern from this repository, we should investigate.

That said, I know this review has been going slowly, so I'm fine with proceeding as-is and following up later. It **should** be an easy cleanup.